R/derive_params_growth_age.R

```diff
@@ -0,0 +1,235 @@
+#'   cdc_meta_criteria,
+#'   who_meta_criteria
+#' ) %>%
+#'   rename(AGEU = age_unit)
+#'
+#' derive_params_growth_age(
+#'   advs,
+#'   sex = SEX,
+#'   age = AGECUR,
+#'   age_unit = AGECURU,
+#'   meta_criteria = criteria,
+#'   parameter = VSTESTCD == "WEIGHT",
+#'   set_values_to_sds = exprs(
+#'     PARAMCD = "WTASDS",
+#'     PARAM = "Weight-for-age z-score"
+#'   ),
+#'   set_values_to_pctl = exprs(
+#'     PARAMCD = "WTAPCTL",
+#'     PARAM = "Weight-for-age percentile"
+#'   )
+#' )
+derive_params_growth_age <- function(dataset,
+                                     sex,
+                                     age,
+                                     age_unit,
+                                     meta_criteria,
+                                     parameter,
+                                     set_values_to_sds = NULL,
+                                     set_values_to_pctl = NULL) {
+  # Apply assertions to each argument to ensure each object is appropriate class
+  # need to add assertion for dataset that checks columns in sex, age, age_unit are present
+  assert_data_frame(meta_criteria, required_vars = exprs(SEX, AGE, AGEU, L, M, S))
+  assert_expr(enexpr(parameter))
+  assert_varval_list(set_values_to_sds, optional = TRUE)
+  assert_varval_list(set_values_to_pctl, optional = TRUE)
+
+  # create a unified join naming convention, hard to figure out in by argument
+  dataset <- dataset %>%
+    mutate(
+      sex_join := {{ sex }},
+      ageu_join := {{ age_unit }}
+    )
+
+  # Process metadata
+  # Metadata should contain SEX, AGE, AGEU, L, M, S
+  # Processing the data to be compatible with the dataset object
+  processed_md <- meta_criteria %>%
+    arrange(SEX, AGEU, AGE) %>%
+    group_by(SEX, AGEU) %>%
+    mutate(next_age = lead(AGE)) %>% # needed for the join and filter later creates [x, y) range
+    rename(
+      sex_join = SEX,
+      prev_age = AGE,
+      ageu_join = AGEU
+    )
+
+  # Merge the dataset that contains the vs records and filter the L/M/S that fit the appropriate age
+  # To parse out the appropriate age, create [x, y) brackets using a prev_age <= AGECUR < next_age
+  added_records <- dataset %>%
+    filter(!!enexpr(parameter)) %>%
+    left_join(.,
+      processed_md,
+      by = c("sex_join", "ageu_join"),
+      relationship = "many-to-many"
+    ) %>%
+    filter(prev_age <= {{ age }} & {{ age }} < next_age)
+
+  dataset_final <- dataset
+
+  # create separate records objects as appropriate depending if user specific sds and/or pctl
+  if (!is_empty(set_values_to_sds)) {
+    add_sds <- added_records %>%
+      mutate(
+        AVAL = ((VSSTRESN / M)^L - 1) / (L * S),
+        !!!set_values_to_sds
+      )
+    dataset_final <- bind_rows(dataset, add_sds) %>%
+      select(-c(L, M, S, sex_join, ageu_join, prev_age, next_age))
+  }
+
+  if (!is_empty(set_values_to_pctl)) {
+    add_pctl <- added_records %>%
+      mutate(
+        AVAL = ((VSSTRESN / M)^L - 1) / (L * S),
```

Comment:
`VSSTRESN` is always present in ADVS but perhaps `AVAL` should be used here? is there a need to double-check that numeric value and unit associated with it is valid/compatible with the metadata?